File: go/2231.go
package main

import "container/heap"

type Heap []int

// Len implements heap.Interface
func (h *Heap) Len() int {
	return len(*h)
}

// Less implements heap.Interface
func (h *Heap) Less(i int, j int) bool {
	return (*h)[i] < (*h)[j]
}

// Swap implements heap.Interface
func (h *Heap) Swap(i int, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}

// Pop implements heap.Interface
func (h *Heap) Pop() interface{} {
  ret := (*h)[len(*h)-1]
  *h = (*h)[:len(*h)-1]
  return ret
}

// Push implements heap.Interface
func (h *Heap) Push(x interface{}) {
  *h = append(*h, x.(int))
}

func largestInteger(num int) int {
  // init min heap for odd and even nums
	var ho, he *Heap = &Heap{}, &Heap{}
	heap.Init(ho)
	heap.Init(he)

	// scan all number and categorize them
	var tmp, cur int = num, 0
	for tmp > 0 {
	  cur = tmp % 10
	  if cur % 2 == 1 {
	    heap.Push(ho, cur)
	  } else {
	    heap.Push(he, cur)
	  }
	  tmp /= 10
	}

	// put them into ret
	var ret, w = 0, 1
	for true {
	  cur = num % 10
	  if cur % 2 == 1 {
	    ret += heap.Pop(ho).(int) * w
	  } else {
	    ret += heap.Pop(he).(int) * w
	  }
	  num /= 10
	  if num > 0 {
	    w *= 10
	  } else {
	    break
	  }
	}
	return ret
}

File: cpp/2231.cpp
#include <functional>
#include <queue>

using namespace std;

class Solution {
public:
  int largestInteger(int num) {
    // min heap
    priority_queue<int, vector<int>, greater<int>> oddheap, evenheap;
    // record all nums and categorize them
    int tmp = num, cur;
    while (tmp > 0) {
      cur = tmp % 10;
      if (cur % 2)
        oddheap.push(cur);
      else
        evenheap.push(cur);
      tmp /= 10;
    }

    // add them back to ret
    int ret = 0, weight = 1;
    while (true) {
      cur = num % 10;
      if (cur % 2) {
        ret += oddheap.top() * weight;
        oddheap.pop();
      } else {
        ret += evenheap.top() * weight;
        evenheap.pop();
      }
      num /= 10;
      if (num > 0)
        weight *= 10;
      else
        break;
    }
    return ret;
  }
};
